File: app/src/main/java/in/digistorm/aksharam/activities/initialise/InitialiseAppActivity.java
package in.digistorm.aksharam.activities.initialise;

/*
 * Copyright (c) 2022 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.json.JSONArray;

import java.io.IOException;

import in.digistorm.aksharam.R;
import in.digistorm.aksharam.activities.main.MainActivity;
import in.digistorm.aksharam.util.GlobalSettings;
import in.digistorm.aksharam.util.LangDataReader;
import in.digistorm.aksharam.util.LanguageDataDownloader;
import in.digistorm.aksharam.util.Log;
import in.digistorm.aksharam.util.OnRequestCompleted;

public class InitialiseAppActivity extends AppCompatActivity {
    private final String logTag = "InitialiseAppActivity";

    private LanguageDataDownloader languageDataDownloader;
    private LanguageDataFileListAdapter adapter;

    private AlertDialog dialog;

    private void showLanguageDataSelectionList(JSONArray languageDataFiles) {
        InitialiseAppActivity self = this;

        // This is called from the background thread; views can be updated only in the threads that
        // created them
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Log.d(logTag, "Updating RecyclerView in UI thread");
                // First, hide the progress bar
                findViewById(R.id.InitialiseAppProgressBar).setVisibility(View.INVISIBLE);
                // Set the appropriate hint
                ((TextView) findViewById(R.id.InitialiseAppHintTV)).setText(R.string.initialisation_choice_hint);

                RecyclerView languageListRV = findViewById(R.id.InitialiseAppLangListRV);
                adapter = new LanguageDataFileListAdapter(languageDataFiles);
                languageListRV.setLayoutManager(new LinearLayoutManager(self));
                languageListRV.setAdapter(adapter);

                ((Button) findViewById(R.id.InitialiseAppProceedButton)).setEnabled(true);
            }
        });
    }

    private boolean startMainActivity() {
        Log.d(logTag, "Attempting to start main activity...");
        // First check if we have data files available
        if(LangDataReader.areDataFilesAvailable(this) == null) {
            Log.d(logTag, "No data files found!! Continuing initialisation activity");
            // we cannot start the main activity
            return false;
        }

        Log.d(logTag, "Starting MainActivity...");
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        // Main activity is started
        return true;
    }

    public void proceed(View v) {
        Log.d(logTag, "Proceed button clicked!");

        if(adapter == null || adapter.getItemCount() <= 0)
            return ;

        JSONArray dataFileList = adapter.getDataFileList();
        for(int i = 0; i < dataFileList.length(); i++) {
            if(dataFileList.optJSONObject(i).optBoolean("selected", false))
                break;

            // if we have gone through all items in the list and not found a single item marked "selected"
            if(i == dataFileList.length() - 1) {
                // show a message
                ((TextView) findViewById(R.id.InitialiseAppHintTV))
                        .setText(R.string.initialisation_no_file_selected);
                // and return
                return ;
            }
        }

        v.setEnabled(false);
        findViewById(R.id.InitialiseAppProgressBar).setVisibility(View.VISIBLE);
        languageDataDownloader.download(adapter.getDataFileList(),
                this,
                new OnRequestCompleted() {
                    @Override
                    public void onDownloadCompleted() {
                        Log.d(logTag, "Download completed; starting MainActivity...");
                        startMainActivity();
                        Log.d(logTag, "InitialiseAppActivity finishing...");
                        finish();
                    }

                    @Override
                    public void onDownloadFailed(Exception e) {
                        v.setEnabled(true);
                        findViewById(R.id.InitialiseAppProgressBar).setVisibility(View.INVISIBLE);
                        Toast.makeText(getApplicationContext(), R.string.could_not_download_file,
                                Toast.LENGTH_LONG).show();
                        Log.d(logTag, "Download failed due to exception " + e);
                        e.printStackTrace();
                    }
                });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(logTag, "Activity destroying...");
    }

    public void showNoInternetDialog() {
        Log.d("NoInternetDialog", "Showing NoInternetDialog.");
        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(this);
        dialogBuilder.setMessage(R.string.could_not_download_file)
                .setTitle(R.string.no_internet)
                .setCancelable(false)
                .setNegativeButton(R.string.exit, (dialog, which) -> {
                    Log.d("NoInternetDialog", "Exit button was clicked");
                    dialog.dismiss();
                    finish();
                })
                .setPositiveButton(R.string.retry_download, (dialog, which) -> {
                    setUpActivity();
                })
                .setOnDismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        Log.d("NoInternetDialog", "Dialog dismissed!");
                    }
                });
        dialog = dialogBuilder.create();
        dialog.show();
    }

    protected void setUpActivity() {
        // Attempt to start main activity
        boolean mainActivityStarted = startMainActivity();
        InitialiseAppActivity activity = this;
        if(!mainActivityStarted) {
            // if we are not able to start the main activity...
            // continue setting up the initialisation activity (current activity)
            setContentView(R.layout.initialise_app_activity);

            languageDataDownloader = new LanguageDataDownloader();

            GlobalSettings.getInstance().getThreadPoolExecutor().execute(() -> {
                try {
                    showLanguageDataSelectionList(languageDataDownloader.getLanguageDataFiles());
                }
                catch (IOException ie) {
                    Log.d(logTag, "IOException caught while downloading language list");
                    runOnUiThread(this::showNoInternetDialog);
                }
            });
            findViewById(R.id.InitialiseAppProceedButton).setOnClickListener(v -> proceed(v));
        }
        else
            finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Log.d(logTag, "Starting app...");
        GlobalSettings.createInstance(this);

        if(GlobalSettings.getInstance().getDarkMode()) {
            // which mode did the activity start in?
            int nightMode = AppCompatDelegate.getDefaultNightMode();
            // If dark mode is enabled, this causes activity to restart
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
            // continue activity initialisation only if activity started in night mode
            if(nightMode == AppCompatDelegate.MODE_NIGHT_YES)
                setUpActivity();
        }
        else { // if light mode; nothing special, it should be light mode by default
            setUpActivity();
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

}

File: app/src/main/java/in/digistorm/aksharam/activities/initialise/InitialiseAppActivity.kt
/*
 * Copyright (c) 2022 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package `in`.digistorm.aksharam.activities.initialise

import `in`.digistorm.aksharam.R
import `in`.digistorm.aksharam.activities.main.MainActivity
import `in`.digistorm.aksharam.util.*

import androidx.appcompat.app.AppCompatActivity
import org.json.JSONArray
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.LinearLayoutManager
import android.content.Intent
import android.widget.Toast
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import java.io.IOException
import java.lang.Exception

class InitialiseAppActivity : AppCompatActivity() {
    private val logTag = InitialiseAppActivity::class.simpleName
    private val languageDataDownloader: LanguageDataDownloader = LanguageDataDownloader()
    private var adapter: LanguageDataFileListAdapter? = null
    private fun showLanguageDataSelectionList(languageDataFiles: JSONArray) {
        val self = this

        // This is called from the background thread; views can be updated only in the threads that
        // created them
        runOnUiThread {
            logDebug(logTag, "Updating RecyclerView in UI thread")
            // First, hide the progress bar
            findViewById<View>(R.id.InitialiseAppProgressBar).visibility = View.INVISIBLE
            // Set the appropriate hint
            (findViewById<View>(R.id.InitialiseAppHintTV) as TextView).setText(R.string.initialisation_choice_hint)
            val languageListRV = findViewById<RecyclerView>(R.id.InitialiseAppLangListRV)
            adapter = LanguageDataFileListAdapter(languageDataFiles)
            languageListRV.layoutManager = LinearLayoutManager(self)
            languageListRV.adapter = adapter
            (findViewById<View>(R.id.InitialiseAppProceedButton) as Button).isEnabled = true
        }
    }

    private fun startMainActivity(): Boolean {
        logDebug(logTag, "Attempting to start main activity...")
        // First check if we have data files available
        if (getDownloadedLanguages(this).isEmpty()) {
            logDebug(logTag, "No data files found!! Continuing initialisation activity")
            // we cannot start the main activity
            return false
        }
        logDebug(logTag, "Starting MainActivity...")
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        // Main activity is started
        return true
    }

    private fun proceed(v: View) {
        logDebug(logTag, "Proceed button clicked!")
        if (adapter == null || adapter!!.itemCount <= 0) return
        val dataFileList = adapter?.dataFileList
        if(dataFileList == null || dataFileList.length() == 0) return

        for (i in 0 until dataFileList.length()) {
            // If at least one file is selected, continue
            if (dataFileList.optJSONObject(i).optBoolean("selected", false))
                break
            // if we have gone through all items in the list and not found a single item marked "selected"
            if (i == dataFileList.length() - 1) {
                // show a message
                (findViewById<View>(R.id.InitialiseAppHintTV) as TextView)
                        .setText(R.string.initialisation_no_file_selected)
                // and return
                return
            }
        }
        v.isEnabled = false
        findViewById<View>(R.id.InitialiseAppProgressBar).visibility = View.VISIBLE
        languageDataDownloader.download(adapter?.dataFileList!!,
            this,
            object : OnRequestCompleted {
                override fun onDownloadCompleted() {
                    logDebug(logTag, "Download completed; starting MainActivity...")
                    startMainActivity()
                    logDebug(logTag, "InitialiseAppActivity finishing...")
                    finish()
                }

                override fun onDownloadFailed(e: Exception?) {
                    v.isEnabled = true
                    findViewById<View>(R.id.InitialiseAppProgressBar).visibility = View.INVISIBLE
                    Toast.makeText(applicationContext, R.string.could_not_download_file,
                        Toast.LENGTH_LONG).show()
                    logDebug(logTag, "Download failed due to exception $e")
                    e?.printStackTrace()
                }
            })
    }

    override fun onDestroy() {
        super.onDestroy()
        logDebug(logTag, "Activity destroying...")
    }

    private fun showNoInternetDialog() {
        logDebug("NoInternetDialog", "Showing NoInternetDialog.")
        val dialogBuilder = AlertDialog.Builder(this)
        dialogBuilder.setMessage(R.string.could_not_download_file)
            .setTitle(R.string.no_internet)
            .setCancelable(false)
            .setNegativeButton(R.string.exit) { dialog: DialogInterface, _: Int ->
                logDebug("NoInternetDialog", "Exit button was clicked")
                dialog.dismiss()
                finish()
            }
            .setPositiveButton(R.string.retry_download) { _: DialogInterface?, _: Int -> setUpActivity() }
            .setOnDismissListener {
                logDebug(
                    "NoInternetDialog",
                    "Dialog dismissed!"
                )
            }
        val dialog1 = dialogBuilder.create()
        dialog1.show()
    }

    private fun setUpActivity() {
        // Attempt to start main activity
        val mainActivityStarted = startMainActivity()
        if (!mainActivityStarted) {
            // if we are not able to start the main activity...
            // continue setting up the initialisation activity (current activity)
            setContentView(R.layout.initialise_app_activity)
            GlobalSettings.instance?.threadPoolExecutor?.execute {
                try {
                    showLanguageDataSelectionList(languageDataDownloader.languageDataFiles)
                } catch (ie: IOException) {
                    logDebug(logTag, "IOException caught while downloading language list")
                    runOnUiThread { showNoInternetDialog() }
                }
            }
            findViewById<View>(R.id.InitialiseAppProceedButton).setOnClickListener { v: View -> proceed(v) }
        } else finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        logDebug(logTag, "Starting app...")
        GlobalSettings.createInstance(this)
        if (GlobalSettings.instance?.darkMode == true) {
            // which mode did the activity start in?
            val nightMode = AppCompatDelegate.getDefaultNightMode()
            // If dark mode is enabled, this causes activity to restart
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            // continue activity initialisation only if activity started in night mode
            if (nightMode == AppCompatDelegate.MODE_NIGHT_YES) setUpActivity()
        } else { // if light mode; nothing special, it should be light mode by default
            setUpActivity()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }
}